### PR TITLE
memtx benchmark

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
# box: fix tuple format and access subsystems initialization

The tuple format and access subsystems have static variables holding their
states which don't get reset during cleanup: initialize them explicitly in
`*_init` functions — that way we can re-initialize these subsystems
multiple times (e.g., when setting up and tearing down benchmarks). Opted
for initializing them in `*_init` functions rather than resetting them in
`*_free` functions for logical consistency.

Needed for #6964

# perf: add memtx benchmark

This first version is quite basic and only benchmarks random `get`s of
existing keys and `select`s of all keys for a tree index (these benchmarks
are needed for #6964) — its main goal is to provide a foundation (i.e., all
the necessary initialization logic) for benchmarking memtx. Extending this
benchmark using the provided memtx singleton and fixture should be fairly
simple.

The results of running this benchmark compiled with clang-16 on my Intel
MacBook Pro (13-inch, 2020) laptop[^1]:

```console
georgiy.lebedev@georgiy-lebedev perf % ./memtx.perftest --benchmark_min_warmup_time=10 --benchmark_repetitions=10 --benchmark_report_aggregates_only=true --benchmark_display_aggregates_only=true
2023-10-02T12:59:36+03:00
Running ./memtx.perftest
Run on (8 X 2000 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB
  L1 Instruction 32 KiB
  L2 Unified 512 KiB (x4)
  L3 Unified 6144 KiB
Load Average: 5.67, 10.05, 7.89
mapping 4398046511104 bytes for memtx tuple arena...
Actual slab_alloc_factor calculated on the basis of desired slab_alloc_factor = 1.090508
fiber has not yielded for more than 0.500 seconds
--------------------------------------------------------------------------------------------------------
Benchmark                                              Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------------------
MemtxFixture/TreeGetRandomExistingKeys_mean          682 ns          667 ns           10 items_per_second=1.51504M/s
MemtxFixture/TreeGetRandomExistingKeys_median        704 ns          693 ns           10 items_per_second=1.44387M/s
MemtxFixture/TreeGetRandomExistingKeys_stddev       81.7 ns         72.7 ns           10 items_per_second=169.696k/s
MemtxFixture/TreeGetRandomExistingKeys_cv          11.97 %         10.90 %            10 items_per_second=11.20%
MemtxFixture/TreeGet1RandomExistingKey_mean          253 ns          241 ns           10 items_per_second=4.20104M/s
MemtxFixture/TreeGet1RandomExistingKey_median        233 ns          229 ns           10 items_per_second=4.36911M/s
MemtxFixture/TreeGet1RandomExistingKey_stddev       46.7 ns         29.7 ns           10 items_per_second=464.187k/s
MemtxFixture/TreeGet1RandomExistingKey_cv          18.43 %         12.34 %            10 items_per_second=11.05%
MemtxFixture/TreeSelectAll_mean                  4766728 ns      4705622 ns           10 items_per_second=27.978M/s
MemtxFixture/TreeSelectAll_median                4605936 ns      4580478 ns           10 items_per_second=28.6184M/s
MemtxFixture/TreeSelectAll_stddev                 447495 ns       349499 ns           10 items_per_second=1.85573M/s
MemtxFixture/TreeSelectAll_cv                       9.39 %          7.43 %            10 items_per_second=6.63%
```

[^1]: https://support.apple.com/kb/SP819?locale=en_US

Needed for #6964

### CODE REVIEW

File: perf/memtx.cc [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+BENCHMARK_F(MemtxFixture, TreeGetRandomExisting)
+(benchmark::State &state)
+{
+	std::int64_t iterations_count = 0;
+	for (auto _: state) {
+		for (auto &k: key_subset) {
+			++iterations_count;
+			struct tuple *t;
+			::box_index_get(sid, tree_iid, k.first, k.second, &t);
```
Comment (alyapunov) [inline perf/memtx.cc:163]: I think `DoNotOptimize` must be added here. Or better check the return value and panic if it's non-zero.

**Review by alyapunov — CHANGES_REQUESTED** [review]

File: perf/memtx.cc (hunk 1 above)
Comment (CuriousGeorgiy) [reply perf/memtx.cc]: Fixed: wrapped it with `DoNotOptimize`.

I believe checking the return value would affect the measurement.

File: src/box/user.cc [hunk 2]
```diff
@@ -610,6 +610,7 @@ user_cache_free(void)
 {
 	if (user_registry)
 		mh_i32ptr_delete(user_registry);
+	memset(users, 0, sizeof(users));
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/user.cc]: Fixed: moved it to `user_cache_init`.

File: src/box/tuple_format.c [hunk 3]
```diff
@@ -1146,6 +1146,11 @@ tuple_format_free()
 		}
 	}
 	free(tuple_formats);
+	tuple_formats = NULL;
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/tuple_format.c]: Fixed: moved it to `tuple_format_init`. Originally did it this way, thinking that these variables are statically initialized to zero.

I don’t quite understand the point of trashing memory in the age of _valgrind_ and _memory sanitizers_.

**Review by Korablev77 — COMMENTED** [review]
>reset them in `*_free` functions

However in patch you reset them in _init functions. Please clarify

**Review by Korablev77 — COMMENTED** [review]
Add a decent comment explaining what happens in benchmark

File: perf/memtx.cc [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+BENCHMARK_F(MemtxFixture, TreeGetRandomExisting)
+(benchmark::State &state)
+{
+	std::int64_t iterations_count = 0;
+	for (auto _: state) {
+		for (const auto &k: key_subset) {
+			++iterations_count;
+			struct tuple *t;
+			benchmark::DoNotOptimize(::box_index_get(sid, tree_iid,
```
Comment (Korablev77) [inline perf/memtx.cc:172]: I guess we should somehow make sure that index_get is proceeded without errors

**Review by Korablev77 — COMMENTED** [review]
Don't get why can't you simply "mock" this function as it is done in `core_test_utils.c`

**Review by Korablev77 — COMMENTED** [review]
Add a decent comment explaining what happens in benchmark.

Also it fails to build:
```
/home/nikita/korablev/tarantool/perf/memtx.cc: In member function ‘virtual void MemtxFixture::SetUp(const benchmark::State&)’:
/home/nikita/korablev/tarantool/perf/memtx.cc:69:15: error: ‘strlen’ is not a member of ‘std’
          std::strlen("perf"), "memtx",
               ^~~~~~
/home/nikita/korablev/tarantool/perf/memtx.cc:69:15: note: suggested alternative:
In file included from /home/nikita/korablev/tarantool/src/lib/small/include/small/mempool.h:37:0,
                 from /home/nikita/korablev/tarantool/src/lib/small/include/small/small.h:34,
                 from /home/nikita/korablev/tarantool/src/box/allocator.h:34,
                 from /home/nikita/korablev/tarantool/perf/memtx.cc:3:
/usr/include/string.h:384:15: note:   ‘strlen’
 extern size_t strlen (const char *__s)

```

**Comment by CuriousGeorgiy** [comment]
> However in patch you reset them in _init functions. Please clarify

I changed it after @alyapunov's comments (https://github.com/tarantool/tarantool/pull/7083#discussion_r858909965 and https://github.com/tarantool/tarantool/pull/7083#discussion_r859126887).

> Add a decent comment explaining what happens in benchmark

Done.

> Don't get why can't you simply "mock" this function as it is done in core_test_utils.c

Originally, I hesitated to pull in the Lua dependency:
https://github.com/tarantool/tarantool/blob/3240201a2f5bac3bddf8a74015db9b351954e0b5/src/lua/utils.c#L1061-L1063

But in the end, it turned out to be a matter of initialing `tarantool_L` with a `luaL_newstate()` in the memtx singleton, so I dropped this commit.

> Also it fails to build:

Fixed. Tested build with clang-16 and gcc-13.

File: perf/memtx.cc (hunk 4 above)
Comment (CuriousGeorgiy) [reply perf/memtx.cc]: I guess this can be checked once manually. On the other hand, adding a check probably won't harm. @alyapunov what do you think?

**Comment by CuriousGeorgiy** [comment]
@locker `lint/checkpatch` seems to give a lot of FPs on C++ code. I think we discussed this a while ago and decided to remove stop remove the `perf` folder from the scope of its checks (perhaps, in the future, we could add a `.clang-format` file instead.

**Comment by CuriousGeorgiy** [comment]
@alyapunov just to note, technically, since I cannot completely cleanup all Tarantool subsystems, I can drop the first commit.

File: perf/memtx.cc [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+/**
+ * Benchmark random `get`s of existing keys from the tree index. The key subset
+ * is regenerated through every iteration. This benchmark is supposed to
+ * exercise the CPU and the cache.
+ */
+BENCHMARK_F(MemtxFixture, TreeGetRandomExistingKeys)
+(benchmark::State &state)
+{
+	struct tuple *t;
```
Comment (alyapunov) [inline perf/memtx.cc:246]: I would make it in the following way:
```
auto itr = key_subset.begin();
for (MAYBE_UNUSED auto _ : state) {
    if (itr == key_subset.end()) {
        state.PauseTiming();
        generate_key_subset();
        state.ResumeTiming();
        itr  = key_subset.begin();
    }
```

File: perf/memtx.cc [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+		rmean_box =
+			::rmean_new(iproto_type_strs, IPROTO_TYPE_STAT_MAX);
+		if (rmean_box == nullptr)
+			panic("failed to create new rmean");
+		rmean_error =
+			::rmean_new(rmean_error_strings, RMEAN_ERROR_LAST);
+		if (rmean_error == nullptr)
+			panic("failed to create new rmean");
+
+		memtx = ::memtx_engine_new(".", true, QUOTA_MAX,
```
Comment (Gumix) [inline perf/memtx.cc:86]: I guess `QUOTA_MAX` is too big for `tuple_arena_max_size`, I get:
```
$ ./perf/memtx.perftest
mapping 4398046511104 bytes for memtx tuple arena...
failed to preallocate 4398046511104 bytes: Cannot allocate memory, check option 'memtx_memory' in box.cfg(..)
```

File: perf/memtx.cc [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+		for (const auto &k: key_subset) {
+			benchmark::DoNotOptimize(::box_index_get(
+				Memtx::sid, Memtx::tree_iid, k.first, k.second,
+				&t));
+		}
+		state.PauseTiming();
+		generate_key_subset();
+		state.ResumeTiming();
+	}
+	state.SetItemsProcessed(key_subset_size);
```
Comment (Gumix) [inline perf/memtx.cc:257]: IIANM, `state.SetItemsProcessed()` value should be calculated this way:
```diff
--- a/perf/memtx.cc
+++ b/perf/memtx.cc
@@ -247,17 +247,19 @@ BENCHMARK_F(MemtxFixture, TreeGetRandomExistingKeys)
 (benchmark::State &state)
 {
        struct tuple *t;
+       size_t total_count = 0;
        for (MAYBE_UNUSED auto _ : state) {
                for (const auto &k: key_subset) {
                        benchmark::DoNotOptimize(::box_index_get(
                                Memtx::sid, Memtx::tree_iid, k.first, k.second,
                                &t));
+                       total_count++;
                }
                state.PauseTiming();
                generate_key_subset();
                state.ResumeTiming();
        }
-       state.SetItemsProcessed(key_subset_size);
+       state.SetItemsProcessed(total_count);
 }
```
Before this patch: `items_per_second=91.8609k/s`
After this patch: `items_per_second=1.53542M/s`

File: perf/memtx.cc (hunk 5 above)
Comment (CuriousGeorgiy) [reply perf/memtx.cc:299]: Fixed.

File: perf/memtx.cc (hunk 7 above)
Comment (CuriousGeorgiy) [reply perf/memtx.cc]: Yeah, I made a typo in all `state.SetItemsProcessed` calls, fixed.

File: perf/memtx.cc [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+	const char *packed_pos = NULL;
+	const char *packed_pos_end = NULL;
+	for (MAYBE_UNUSED auto _ : state) {
+		/* Vary iterator type to reduce cache impact. */
+		enum iterator_type iter_type = iter_parity ? ITER_ALL : ITER_LE;
+		benchmark::DoNotOptimize(::box_select(
+			Memtx::sid, Memtx::tree_iid, iter_type, 0, select_lim,
+			empty_key.first, empty_key.second, &packed_pos,
+			&packed_pos_end, false, &port));
+		iter_parity += 1;
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply perf/memtx.cc]: We discussed this with @alyapunov, and he said this iterator type change looks redundant (initially I wanted to mitigate some potential caching and branch prediction effects), so I dropped it.

File: perf/memtx.cc (hunk 6 above)
Comment (CuriousGeorgiy) [reply perf/memtx.cc]: Didn't notice it on my laptop, sorry. Changed it to `1 GiB`, I guess one can tweak it further if necessary.

**Review by Gumix — APPROVED** [review]
The benchmark is actually not compiled after https://github.com/tarantool/tarantool/commit/686233814d314de7600fcf33cfbac585bb08cc46. Here is the fix:
```diff
--- a/perf/CMakeLists.txt
+++ b/perf/CMakeLists.txt
@@ -77,6 +77,13 @@ create_perf_test_target(TARGET light)

 create_perf_test_target(TARGET small)

+create_perf_test(NAME memtx
+                 SOURCES memtx.cc ${PROJECT_SOURCE_DIR}/test/unit/box_test_utils.c
+                 LIBRARIES core box server benchmark::benchmark
+)
+
+create_perf_test_target(TARGET memtx)
+
 add_custom_target(test-c-perf
                   DEPENDS ${RUN_PERF_C_TESTS_LIST}
                   COMMENT "Running C performance tests"
@@ -86,8 +93,3 @@ add_custom_target(test-perf
                   DEPENDS test-c-perf test-lua-perf
                   COMMENT "Running performance tests"
 )
-
-create_perf_test(PREFIX memtx
-                 SOURCES memtx.cc ${PROJECT_SOURCE_DIR}/test/unit/box_test_utils.c
-                 LIBRARIES core box server benchmark::benchmark
-)
```

**Comment by CuriousGeorgiy** [comment]
@Gumix didn't notice — fixed that problem, thanks!

File: perf/CMakeLists.txt [hunk 9]
```diff
@@ -77,6 +77,12 @@ create_perf_test_target(TARGET light)
 
 create_perf_test_target(TARGET small)
 
+create_perf_test(PREFIX memtx
```
Comment (Gumix) [inline perf/CMakeLists.txt:80]: `:s/PREFIX/NAME/`

Comment (CuriousGeorgiy) [reply perf/CMakeLists.txt]: Fixed, I am sorry 🤦‍♂️.

**Review by alyapunov — APPROVED** [review]

File: perf/memtx.cc (hunk 4 above)
Comment (alyapunov) [reply perf/memtx.cc]: That would be good but I think not necessary. I bet in case of errors or nops we will explicitly notice it by bench results.

**Pull request merged**